File: ios/PulseLM/PulseLM/Mapping/RangeLanding.swift
import Foundation

/// Landing in yards relative to the target line. Matches range_landing.py.
struct RangeLanding: Equatable, Sendable {
    var alongYd: Double?
    var offlineYd: Double?
    /// True when HLA is missing (including missing carry with null HLA).
    var onLine: Bool

    /// Matches `range_landing.landing_yd`:
    /// - If `carry` is nil: along/offline nil, onLine = (hla == nil).
    /// - If `hla` is nil: along = carry, offline = 0, onLine true.
    /// - Else: along = carry * cos(hla rad), offline = carry * sin(hla rad), onLine false.
    /// Positive offline is right of the target line.
    static func from(carry: Double?, hla: Double?) -> RangeLanding {
        guard let carry else {
            return RangeLanding(alongYd: nil, offlineYd: nil, onLine: hla == nil)
        }
        guard let hla else {
            return RangeLanding(alongYd: carry, offlineYd: 0, onLine: true)
        }
        let radians = hla * Double.pi / 180.0
        return RangeLanding(
            alongYd: carry * cos(radians),
            offlineYd: carry * sin(radians),
            onLine: false
        )
    }

    static func from(shot: ShotResult) -> RangeLanding {
        from(carry: shot.carry_yd_est, hla: shot.hla_deg)
    }
}

File: range_landing.py
"""Project carry onto range along/offline yards. HLA stays null unless measured."""

from __future__ import annotations

from math import cos, radians, sin
from typing import Any, Optional


def landing_yd(
    carry_yd_est: Optional[float], hla_deg: Optional[float] = None
) -> dict[str, Any]:
    """along = carry if HLA is missing; else carry * cos/sin(HLA)."""
    if carry_yd_est is None:
        return {
            "along_yd": None,
            "offline_yd": None,
            "on_line": hla_deg is None,
        }
    carry = float(carry_yd_est)
    if hla_deg is None:
        return {
            "along_yd": carry,
            "offline_yd": 0.0,
            "on_line": True,
        }
    hla_rad = radians(float(hla_deg))
    return {
        "along_yd": carry * cos(hla_rad),
        "offline_yd": carry * sin(hla_rad),
        "on_line": False,
    }


def landing_from_shot(shot: dict[str, Any]) -> dict[str, Any]:
    out = landing_yd(shot.get("carry_yd_est"), shot.get("hla_deg"))
    out["shot_id"] = shot.get("shot_id")
    out["carry_yd_est"] = shot.get("carry_yd_est")
    out["hla_deg"] = shot.get("hla_deg")
    return out
